keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1023,26 +1025,146 @@ this enhancement.
+  enforcement point, higher level constructs can map to the actions of this
+  enforcement point as needed.
+
+Future work:
+
+- ValidatingAdmissionPolicy resources might, in the future, add a `warnings`
+  field adjacent to the `validations` and `auditAnnotations` fields to declare
+  expressions only ever result in warnings. This would allow
+  ValidatingAdmissionPolicy authors to declare a expression as non-enforcing
+  regardless of `validationActions`.
+
+- ValidatingAdmissionPolicy resources, might, in the future, offer per-expression
+  enforcement actions (instead of a separate `warnings` field) and combine these
+  enforcement actions with the ValidatingAdmissionPolicyBinding enforcement action
+  to determine the effective enforcement.
+
+#### Audit Annotations
+
+`ValidatingAdmissionPolicy` may declare [Audit
+  annotations](https://github.com/kubernetes/kubernetes/blob/97bbf07d3f3f20332912ee411fdf75ce84425e28/staging/src/k8s.io/api/admission/v1/types.go#L142)
+  in the policy definition. E.g.:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  ...
+  validations:
+    - expression: <expression>
+  auditAnnotations:
+    - key: "my-audit-key"
+      valueExpression: <expression that evaluates to a string (and is recorded) or null (and is not recorded)>
+```
+
+`auditAnnotations` are independent of `validations`. A `ValidatingAdmissionPolicy`
+may contain only `validations`, only `auditAnnotations` or both.
+
+The published annotation key will be of the form `<ValidatingPolicyDefinition
+name>/<auditAnnotation key>` and will be validated as a
+[QualifiedName](https://github.com/kubernetes/kubernetes/blob/dfa4143086bf504c6c72d5eee8a2210b8ed41b9a/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L43).
+
+The validation rule will be: `len(key) < QualifierName.maxLength - len(policy
+name) - 1` to accommodate the `<ValidatingPolicyDefinition
+name>/<auditAnnotation key>` audit annotation key format.
 
-We also intend to support multiple enforcement actions:
+If `valueExpression` returns a string, the audit annotation is published.  If
+`valueExpression` returns null, the audit annotation is omitted. No other return
+types will be supported.
 
-- Deny
-- Audit annotation
-- Client warnings
+#### Audit Events
+
+All audit event keys are prefixed by `<ValidatingPolicyDefinition name>/`.
+
+At Metadata audit level or higher, when a validating admission binding fails any
+validation expression, details are included in the audit annotations
+for the audit event under the key `validation_failures`. E.g.:
+
+```yaml
+# the audit event recorded
+{
+    "kind": "Event",
+    "apiVersion": "audit.k8s.io/v1",
+    "annotations": {
+        "mypolicy.mygroup.example.com/validation_failure": "{\"expression\": 1, \"message\": \"x must be greater than y\", \"enforcement\": \"Deny\", \"binding\": \"mybinding.mygroup.example.com\"}"
```

Comment:
not something you need to address here, but I'm starting to worry that we're overusing audit annotations, and should start moving more metadata like this to dedicated fields.

Reply:
Good point. I'm not satisfied with how the value payload works here. I'll admit to following precedence of other audit annotations rather than trying to re-think it.  That said, I'd be okay with restructuring this. Any recommendations?